File: ambient-explainer-backend/app/factcheck.py
"""
Server-side fact-checking — the write path behind the shared errata cache.

This mirrors the extension's factcheck.js deliberately, rule for rule, because
the two have to agree: a reader who triggers a live check and a reader served
the cached row must see the same verdict for the same sentence, or the cache
looks broken rather than fast.

The evidence rule is unchanged from the extension. A verdict is only worth
showing if it's grounded in something the reader can click, so nothing here ever
asks a model "is this true?" from memory:

  1. PUBLISHED FACT-CHECKS (free) — Google's ClaimReview index. A hit means a
     human fact-checker already ruled. Strongest answer available, costs nothing.
  2. SEARCH + MODEL (paid) — only when step 1 finds nothing. A search-native
     model so retrieval and reasoning happen in one call, keeping the citations.

What differs is who pays. Here the key is the server's, so the cost of a check
is bounded by the number of *articles*, not the number of readers.
"""

import json
import os
import re
from concurrent.futures import ThreadPoolExecutor

import requests

GOOGLE_FACTCHECK_API_KEY = os.getenv("GOOGLE_FACTCHECK_API_KEY")
GOOGLE_FACTCHECK_URL = "https://factchecktools.googleapis.com/v1alpha1/claims:search"

AI_GATEWAY_API_KEY = os.getenv("AI_GATEWAY_API_KEY", "")
AI_GATEWAY_URL = "https://ai-gateway.vercel.sh/v1/chat/completions"

# Perplexity's sonar models search the web as part of ordinary generation and
# return a citations array, so a plain chat/completions POST yields a grounded
# answer with no tool-calling round trip.
SEARCH_MODEL = os.getenv("FACTCHECK_SEARCH_MODEL", "perplexity/sonar")
# Claim extraction is ordinary text work on text we already hold, so it runs on
# something cheap. In the extension this job goes to the on-device model; there
# is no on-device model on a server, so the cheapest hosted one stands in.
EXTRACT_MODEL = os.getenv("FACTCHECK_EXTRACT_MODEL", "openai/gpt-5.4-mini")

VERDICTS = ["TRUE", "MOSTLY_TRUE", "MIXED", "MOSTLY_FALSE", "FALSE", "UNVERIFIABLE"]

# How many claims one article is allowed to spend. The cap is what keeps a
# 6,000-word feature from costing sixty search calls.
MAX_CLAIMS = int(os.getenv("FACTCHECK_MAX_CLAIMS", "6"))

VERDICT_RULES = """
Rules:
- Judge ONLY objectively checkable assertions. Opinions, predictions, promises,
  value judgements and rhetorical questions are UNVERIFIABLE by definition.
- Use UNVERIFIABLE freely. It is the correct answer whenever the evidence you
  found does not actually settle the claim. Never guess to seem useful.
- Judge the claim as stated, not a stronger or weaker version of it.
- Every source URL you cite must be one you actually retrieved."""

EXTRACT_RULES = """Extract only sentences that state a checkable fact — a number, date, event, quantity, attribution or causal assertion that could in principle be shown false.
Skip opinions, predictions, promises, questions, definitions and generic statements."""


# ---------- 1. Published fact-checks (free) ----------

def lookup_published(claim: str, language: str = "en") -> dict:
    """
    Published rulings for a claim, straight from Google's ClaimReview index.

    Returns {"ok": bool, "results": [...]}. `ok=False` means the lookup could
    not run at all, which is NOT the same as running and finding nothing — most
    claims have simply never been formally fact-checked. Collapsing the two
    makes a broken lookup indistinguishable from an unruled claim, and the
    reader can't tell they're being told something false.
    """
    query = (claim or "").strip()[:300]
    if not query:
        return {"ok": True, "results": []}
    if not GOOGLE_FACTCHECK_API_KEY:
        return {"ok": True, "results": []}

    try:
        response = requests.get(
            GOOGLE_FACTCHECK_URL,
            params={
                "query": query,
                "languageCode": language,
                "pageSize": 5,
                "key": GOOGLE_FACTCHECK_API_KEY,
            },
            timeout=8,
        )
    except requests.RequestException:
        return {"ok": False, "results": []}

    if response.status_code != 200:
        print("FACTCHECK LOOKUP FAILURE:", response.status_code, response.text[:200])
        return {"ok": False, "results": []}

    claims = response.json().get("claims", [])
    results = []
    for item in claims:
        for review in item.get("claimReview", []) or []:
            results.append({
                "claim": item.get("text") or query,
                "claimant": item.get("claimant") or "",
                "publisher": (review.get("publisher") or {}).get("name") or "",
                "rating": review.get("textualRating") or "",
                "title": review.get("title") or "",
                "url": review.get("url") or "",
            })
    return {"ok": True, "results": results}


def rating_to_verdict(rating: str) -> str:
    """
    Map a publisher's own wording ("Pants on Fire", "Four Pinocchios",
    "Correct") onto our scale. Deliberately conservative: anything that can't be
    read confidently stays MIXED rather than being forced to a strong verdict.
    """
    r = (rating or "").lower()
    if re.search(r"pants on fire|pinocchio|fabricat|no evidence|baseless|debunk", r):
        return "FALSE"
    if re.search(r"^false|mostly false|incorrect|inaccurate|wrong|misleading", r):
        return "MOSTLY_FALSE" if re.search(r"mostly|partly", r) else "FALSE"
    if re.search(r"half|mixed|partly|partially|lacks context|missing context|exaggerat", r):
        return "MIXED"
    if re.search(r"mostly true|mostly accurate|largely", r):
        return "MOSTLY_TRUE"
    if re.search(r"^true|accurate|correct|confirmed|verified", r):
        return "TRUE"
    return "MIXED"


# ---------- Gateway ----------

def _gateway(model: str, prompt: str, max_tokens: int) -> dict | None:
    if not AI_GATEWAY_API_KEY:
        return None
    try:
        response = requests.post(
            AI_GATEWAY_URL,
            headers={
                "Authorization": f"Bearer {AI_GATEWAY_API_KEY}",
                "Content-Type": "application/json",
            },
            json={
                "model": model,
                "messages": [{"role": "user", "content": prompt}],
                "max_tokens": max_tokens,
                "temperature": 0,
            },
            timeout=60,
        )
    except requests.RequestException as exc:
        print("GATEWAY UNREACHABLE:", exc)
        return None

    if response.status_code != 200:
        print("GATEWAY FAILURE:", response.status_code, response.text[:200])
        return None
    try:
        return response.json()
    except ValueError:
        return None


def _content(data: dict | None) -> str:
    if not data:
        return ""
    try:
        return data["choices"][0]["message"]["content"] or ""
    except (KeyError, IndexError, TypeError):
        return ""


# ---------- 2. Claim extraction ----------

def _extract_prompt(text: str, limit: int) -> str:
    # Two forms per claim, and both are load-bearing. `claim` stands alone so it
    # can be verified out of context; `quote` is the sentence exactly as printed
    # so the page can underline it in place. A paraphrased quote can't be found.
    return f"""{EXTRACT_RULES}
Rank by how much a reader would benefit from having it verified, and return at most {limit}.

For each, return an object with:
- "quote": the sentence copied EXACTLY from the text, character for character (same words, casing and punctuation) so it can be found again. Do not fix or shorten it.
- "claim": the same statement rewritten to stand on its own (resolve "he", "it", "last year" into the names and dates the passage makes clear).

TEXT:
\"\"\"{text[:9000]}\"\"\"

Reply with ONLY a JSON array of objects, no markdown fence:
[{{"quote":"...","claim":"..."}}]"""


def _parse_claims(raw: str, limit: int) -> list:
    match = re.search(r"\[[\s\S]*\]", raw or "")
    if not match:
        return []
    try:
        arr = json.loads(match.group())
    except json.JSONDecodeError:
        return []
    if not isinstance(arr, list):
        return []

    claims = []
    for item in arr:
        if isinstance(item, str):
            text = item.strip()
            claim, quote = text, text
        elif isinstance(item, dict):
            claim = str(item.get("claim") or item.get("quote") or "").strip()
            quote = str(item.get("quote") or item.get("claim") or "").strip()
        else:
            continue
        if len(claim) > 12:
            claims.append({"claim": claim, "quote": quote})
    return claims[:limit]


def extract_claims(text: str, limit: int = MAX_CLAIMS) -> list:
    body = (text or "").strip()
    if len(body) < 40:
        return []
    raw = _content(_gateway(EXTRACT_MODEL, _extract_prompt(body, limit), 700))
    return _parse_claims(raw, limit)


# ---------- 3. Search-grounded verdict ----------

def _verdict_prompt(claim: str, context: str, published: list) -> str:
    priors = ""
    if published:
        lines = "\n".join(
            f'- {p["publisher"]} rated it "{p["rating"]}" — {p["url"]}' for p in published
        )
        priors = f"\nPublished fact-checks already found for this claim (weigh these heavily):\n{lines}\n"

    where = ""
    if context:
        where = (
            "\nIt appeared in this passage, so judge the meaning it carries there:\n"
            f'"""{context[:1200]}"""\n'
        )

    return f"""Fact-check this claim by searching for current evidence.

CLAIM: "{claim}"
{where}{priors}{VERDICT_RULES}

Reply with ONLY a JSON object, no markdown fence, in exactly this shape:
{{"verdict":"{'|'.join(VERDICTS)}","confidence":"high|medium|low","summary":"<=40 words, plain language, what the evidence shows","sources":[{{"title":"...","url":"..."}}]}}"""


def _parse_verdict(raw: str) -> dict | None:
    match = re.search(r"\{[\s\S]*\}", raw or "")
    if not match:
        return None
    try:
        parsed = json.loads(match.group())
    except json.JSONDecodeError:
        return None

    verdict = re.sub(r"[\s-]+", "_", str(parsed.get("verdict") or "").upper())
    sources = []
    if isinstance(parsed.get("sources"), list):
        for s in parsed["sources"][:6]:
            if isinstance(s, dict) and re.match(r"^https?://", str(s.get("url") or "")):
                sources.append({
                    "title": str(s.get("title") or s.get("url"))[:140],
                    "url": s["url"],
                })

    confidence = parsed.get("confidence")
    return {
        "verdict": verdict if verdict in VERDICTS else "UNVERIFIABLE",
        "confidence": confidence if confidence in ("high", "medium", "low") else "low",
        "summary": str(parsed.get("summary") or "").strip(),
        "sources": sources,
    }


def verify(claim: str, context: str, published: list) -> dict | None:
    data = _gateway(SEARCH_MODEL, _verdict_prompt(claim, context, published), 700)
    if data is None:
        return None
    parsed = _parse_verdict(_content(data))
    if not parsed:
        return None

    # Sonar returns the pages it actually read; those beat anything the model
    # typed into the JSON, so merge them in and de-duplicate by URL.
    cited = []
    for url in data.get("citations") or []:
        if isinstance(url, str) and re.match(r"^https?://", url):
            title = re.sub(r"^https?://(www\.)?", "", url).split("/")[0]
            cited.append({"title": title, "url": url})

    seen = set()
    merged = []
    for source in parsed["sources"] + cited:
        if source["url"] in seen:
            continue
        seen.add(source["url"])
        merged.append(source)
    parsed["sources"] = merged[:6]
    return parsed


# ---------- Orchestration ----------

def check_one(claim: str, context: str) -> dict | None:
    """One claim in, one verdict out. Published rulings win when they exist."""
    text = (claim or "").strip()
    if not text:
        return None

    lookup = lookup_published(text)
    published = lookup["results"]

    verified = verify(text, context, published)

    if verified is None:
        # The search leg couldn't run. A published ruling can still stand on its
        # own; otherwise say plainly that the check didn't complete rather than
        # inventing a ruling. Neither of these gets cached — see is_cacheable.
        if published:
            top = published[0]
            return {
                "claim": text,
                "verdict": rating_to_verdict(top["rating"]),
                "confidence": "high",
                "summary": f'{top["publisher"]} rated this "{top["rating"]}".',
                "sources": [
                    {"title": f'{p["publisher"]}{" — " + p["title"] if p["title"] else ""}',
                     "url": p["url"]}
                    for p in published[:4]
                ],
                "origin": "published",
            }
        return {
            "claim": text,
            "verdict": "UNVERIFIABLE",
            "confidence": "low",
            "summary": (
                "Couldn't complete the check — the search step was unavailable."
                if lookup["ok"]
                else "Couldn't reach the fact-check lookup."
            ),
            "sources": [],
            "origin": "error",
        }

    return {
        **verified,
        "claim": text,
        "sources": verified["sources"] or [
            {"title": p["publisher"], "url": p["url"]} for p in published[:3]
        ],
        "origin": "searched+published" if published else "searched",
    }


def check_text(text: str, limit: int = MAX_CLAIMS) -> dict:
    """
    Check a body of text end to end: pull the claims worth checking, then verify
    them concurrently. Each verdict carries the quote it came from so the page
    can tie the ruling back to the exact sentence it underlines.

    Threads rather than async because every leg of this is a blocking `requests`
    call, and claims are independent — six sequential search calls would put a
    minute on a cold article for no reason.
    """
    if not AI_GATEWAY_API_KEY:
        return {"ok": False, "reason": "no-gateway-key", "claims": [], "verdicts": []}

    claims = extract_claims(text, limit)
    if not claims:
        return {"ok": True, "claims": [], "verdicts": []}

    with ThreadPoolExecutor(max_workers=min(6, len(claims))) as pool:
        results = list(pool.map(lambda c: _safe_check(c, text), claims))

    verdicts = []
    for claim, result in zip(claims, results):
        if not result:
            continue
        result["quote"] = claim["quote"]
        verdicts.append(result)

    return {"ok": True, "claims": claims, "verdicts": verdicts}


def _safe_check(claim: dict, context: str) -> dict | None:
    try:
        return check_one(claim["claim"], context)
    except Exception as exc:  # one bad claim must not sink the article
        print("CLAIM CHECK FAILURE:", exc)
        return None

File: ambient-explainer-extension/factcheck.js
// factcheck.js — evidence-backed fact-checking for JustClarify.
//
// A fact-check is only worth showing if it's grounded in something the reader
// can click. So this never asks a model "is this true?" from memory. Two
// sources of evidence, cheapest first:
//
//   1. PUBLISHED FACT-CHECKS (free). Google's Fact Check Tools API indexes
//      ClaimReview markup from PolitiFact, Snopes, FactCheck.org, AFP and
//      hundreds more. A hit here means a human fact-checker already ruled on
//      this claim — the strongest answer we can give, and it costs nothing.
//   2. WEB SEARCH + MODEL (needs a Gateway key). Only when step 1 finds
//      nothing. We use a search-native model so retrieval and reasoning happen
//      in one call, and we keep the returned citations.
//
// Chrome's on-device model is deliberately NOT used to reach a verdict: Gemini
// Nano can't browse and has a small, dated knowledge base, so it would produce
// confident fiction. It IS used for claim extraction, which is pure text work
// on text we already have — see factCheckExtractClaims.
//
// Loaded by background.js after gateway.js (reuses gatewayGetSettings).

const FC_LOOKUP_PROXY = "https://api.justclarify.xyz/factcheck/lookup";
const FC_GOOGLE_ENDPOINT =
  "https://factchecktools.googleapis.com/v1alpha1/claims:search";

// Perplexity's sonar models search the web as part of ordinary generation and
// return a citations array — which means grounded answers over a plain
// chat/completions POST, with no tool-calling round trip. That matters here:
// this extension talks to the Gateway with raw fetch, not the AI SDK.
const FC_SEARCH_MODEL = "perplexity/sonar";

const FC_VERDICTS = [
  "TRUE",
  "MOSTLY_TRUE",
  "MIXED",
  "MOSTLY_FALSE",
  "FALSE",
  "UNVERIFIABLE",
];

// Claims we refuse to rule on. Fact-checking an opinion is a category error and
// the fastest way to lose a reader's trust.
const FC_VERDICT_RULES = `
Rules:
- Judge ONLY objectively checkable assertions. Opinions, predictions, promises,
  value judgements and rhetorical questions are UNVERIFIABLE by definition.
- Use UNVERIFIABLE freely. It is the correct answer whenever the evidence you
  found does not actually settle the claim. Never guess to seem useful.
- Judge the claim as stated, not a stronger or weaker version of it.
- Every source URL you cite must be one you actually retrieved.`;

// ---------- 1. Published fact-checks (free) ----------

// Prefer the user's own Google API key when they've added one — that path works
// with no JustClarify server at all, which matters for self-hosters. Otherwise
// go through the proxy so the default install needs zero setup.
// Returns { ok, results }. `ok:false` means the lookup could not run at all
// (offline, proxy down, bad key) — which is NOT the same as running and finding
// nothing, and the two need different words in the UI. Collapsing both to an
// empty array made a broken lookup look identical to an unruled claim.
async function factCheckLookup(claim) {
  const query = String(claim || "").trim().slice(0, 300);
  if (!query) return { ok: true, results: [] };

  const { jcGoogleFactKey } = await chrome.storage.local
    .get(["jcGoogleFactKey"])
    .catch(() => ({}));

  const url = jcGoogleFactKey
    ? `${FC_GOOGLE_ENDPOINT}?query=${encodeURIComponent(query)}&languageCode=en&pageSize=5&key=${encodeURIComponent(jcGoogleFactKey)}`
    : `${FC_LOOKUP_PROXY}?query=${encodeURIComponent(query)}`;

  try {
    const res = await fetch(url);
    if (!res.ok) {
      console.warn("JustClarify: fact-check lookup failed", res.status, url.split("?")[0]);
      return { ok: false, results: [] };
    }
    const data = await res.json();
    return {
      ok: true,
      results: (data.claims || []).flatMap((c) =>
        (c.claimReview || []).map((r) => ({
          claim: c.text || query,
          claimant: c.claimant || "",
          publisher: r.publisher?.name || "",
          rating: r.textualRating || "",
          title: r.title || "",
          url: r.url || "",
        })),
      ),
    };
  } catch (error) {
    console.warn("JustClarify: fact-check lookup unreachable", error);
    return { ok: false, results: [] };
  }
}

// Map a publisher's own wording ("Pants on Fire", "Four Pinocchios", "Correct")
// onto our verdict scale. Deliberately conservative: anything we can't read
// confidently stays MIXED rather than being forced to a strong verdict.
function factCheckRatingToVerdict(rating) {
  const r = String(rating || "").toLowerCase();
  if (/pants on fire|pinocchio|fabricat|no evidence|baseless|debunk/.test(r)) return "FALSE";
  if (/^false|mostly false|incorrect|inaccurate|wrong|misleading/.test(r)) {
    return /mostly|partly/.test(r) ? "MOSTLY_FALSE" : "FALSE";
  }
  if (/half|mixed|partly|partially|lacks context|missing context|exaggerat/.test(r)) return "MIXED";
  if (/mostly true|mostly accurate|largely/.test(r)) return "MOSTLY_TRUE";
  if (/^true|accurate|correct|confirmed|verified/.test(r)) return "TRUE";
  return "MIXED";
}

// ---------- 2. Search-grounded verdict (needs a Gateway key) ----------

function factCheckPrompt(claim, context, published) {
  const priors = published.length
    ? `\nPublished fact-checks already found for this claim (weigh these heavily):\n${published
        .map((p) => `- ${p.publisher} rated it "${p.rating}" — ${p.url}`)
        .join("\n")}\n`
    : "";

  const where = context
    ? `\nIt appeared in this passage, so judge the meaning it carries there:\n"""${context.slice(0, 1200)}"""\n`
    : "";

  return `Fact-check this claim by searching for current evidence.

CLAIM: "${claim}"
${where}${priors}${FC_VERDICT_RULES}

Reply with ONLY a JSON object, no markdown fence, in exactly this shape:
{"verdict":"${FC_VERDICTS.join('|')}","confidence":"high|medium|low","summary":"<=40 words, plain language, what the evidence shows","sources":[{"title":"...","url":"..."}]}`;
}

function factCheckParse(raw) {
  const match = String(raw || "").match(/\{[\s\S]*\}/);
  if (!match) return null;
  let parsed;
  try {
    parsed = JSON.parse(match[0]);
  } catch (_) {
    return null;
  }
  const verdict = String(parsed.verdict || "").toUpperCase().replace(/[\s-]+/g, "_");
  return {
    verdict: FC_VERDICTS.includes(verdict) ? verdict : "UNVERIFIABLE",
    confidence: ["high", "medium", "low"].includes(parsed.confidence) ? parsed.confidence : "low",
    summary: String(parsed.summary || "").trim(),
    sources: Array.isArray(parsed.sources)
      ? parsed.sources
          .filter((s) => s && typeof s.url === "string" && /^https?:\/\//.test(s.url))
          .slice(0, 6)
          .map((s) => ({ title: String(s.title || s.url).slice(0, 140), url: s.url }))
      : [],
  };
}

async function factCheckVerify(claim, context, published) {
  const settings = await gatewayGetSettings();

  // No key of their own: JustClarify's hosted tier verifies instead. This used
  // to `return null` here, which meant fact-checking — the feature the product
  // is named after — silently did nothing for every user without an API key.
  if (!settings) {
    const answer = await hostedComplete(
      [{ role: "user", content: factCheckPrompt(claim, context, published) }],
      { maxTokens: 700 },
    );
    if (!answer) return { error: "the fact-check service is unavailable right now" };
    const parsed = factCheckParse(answer);
    // The hosted model has no web search, so it returns no citations. A verdict
    // it can't source is one it shouldn't be trusted to rule on — factCheckParse
    // already carries whatever sources the model named, and the caller shows
    // UNVERIFIABLE when there are none.
    return parsed || { error: "the model's answer couldn't be read as a verdict" };
  }

  let res;
  try {
    res = await fetch("https://ai-gateway.vercel.sh/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${settings.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: FC_SEARCH_MODEL,
        messages: [{ role: "user", content: factCheckPrompt(claim, context, published) }],
        max_tokens: 700,
        temperature: 0,
      }),
    });
  } catch (e) {
    return { error: `couldn't reach the Gateway (${String(e).slice(0, 80)})` };
  }

  if (!res.ok) {
    const body = await res.text().catch(() => "");
    if (res.status === 401 || res.status === 403) {
      return { error: "the Gateway rejected your API key" };
    }
    if (res.status === 402) return { error: "this Gateway key is out of credit" };
    return { error: `Gateway returned HTTP ${res.status}: ${body.slice(0, 120)}` };
  }

  const data = await res.json().catch(() => null);
  const parsed = factCheckParse(data?.choices?.[0]?.message?.content);
  if (!parsed) return { error: "the model's answer couldn't be read as a verdict" };

  // Sonar returns the pages it actually read; those beat anything the model
  // typed into the JSON, so merge them in and de-duplicate by URL.
  const cited = Array.isArray(data.citations)
    ? data.citations
        .filter((u) => typeof u === "string" && /^https?:\/\//.test(u))
        .map((u) => ({ title: u.replace(/^https?:\/\/(www\.)?/, "").split("/")[0], url: u }))
    : [];
  const seen = new Set();
  parsed.sources = [...parsed.sources, ...cited]
    .filter((s) => !seen.has(s.url) && seen.add(s.url))
    .slice(0, 6);

  return parsed;
}

// ---------- Orchestration ----------

// One claim → one verdict. Published fact-checks are authoritative when they
// exist; otherwise we search, and when we can't search we say so plainly
// instead of inventing a ruling.
async function factCheckOne(claim, context) {
  const text = String(claim || "").trim();
  if (!text) return null;

  const lookup = await factCheckLookup(text);
  const published = lookup.results;
  const hasKey = !!(await gatewayGetSettings());

  if (published.length) {
    const top = published[0];
    // With a key we still search — a 2019 ruling on a claim that resurfaced
    // today deserves current context. Without one, the published verdict
    // stands on its own, which is exactly the free tier's promise.
    if (!hasKey) {
      return {
        claim: text,
        verdict: factCheckRatingToVerdict(top.rating),
        confidence: "high",
        summary: `${top.publisher} rated this "${top.rating}".`,
        sources: published.slice(0, 4).map((p) => ({
          title: `${p.publisher}${p.title ? ` — ${p.title}` : ""}`,
          url: p.url,
        })),
        origin: "published",
      };
    }
  }

  if (!hasKey) {
    return {
      claim: text,
      verdict: "UNVERIFIABLE",
      confidence: "low",
      // Two different situations, two different sentences. Saying "nobody has
      // ruled on this" when the lookup never actually ran is a lie the reader
      // can't detect.
      summary: lookup.ok
        ? "No fact-checker has published a ruling on this. The free check covers claims PolitiFact, Snopes, FactCheck.org and others have formally ruled on — mostly political and widely-circulated claims, not ordinary sentences. Add an AI Gateway key in the JustClarify popup to check anything else."
        : "Couldn't reach the fact-check lookup — check your connection and try again.",
      sources: [],
      origin: lookup.ok ? "no-key" : "lookup-failed",
    };
  }

  const verified = await factCheckVerify(text, context, published);
  if (!verified) return null;
  if (verified.error) {
    return {
      claim: text,
      verdict: "UNVERIFIABLE",
      confidence: "low",
      summary: `Couldn't complete the check — ${verified.error}.`,
      sources: published.slice(0, 3).map((p) => ({ title: p.publisher, url: p.url })),
      origin: "error",
    };
  }

  return {
    ...verified,
    claim: text,
    sources: verified.sources.length
      ? verified.sources
      : published.slice(0, 3).map((p) => ({ title: p.publisher, url: p.url })),
    origin: published.length ? "searched+published" : "searched",
  };
}

// ---------- Claim extraction (free on-device when possible) ----------

const FC_EXTRACT_RULES = `Extract only sentences that state a checkable fact — a number, date, event, quantity, attribution or causal assertion that could in principle be shown false.
Skip opinions, predictions, promises, questions, definitions and generic statements.`;

// Each claim carries two forms: `claim` is rewritten to stand alone (pronouns
// and relative dates resolved) so it can be verified out of context; `quote` is
// the exact sentence as it appears in the text so the page can underline it in
// place. The quote MUST be copied verbatim — a paraphrase can't be located.
function factCheckExtractPrompt(text, limit) {
  return `${FC_EXTRACT_RULES}
Rank by how much a reader would benefit from having it verified, and return at most ${limit}.

For each, return an object with:
- "quote": the sentence copied EXACTLY from the text, character for character (same words, casing and punctuation) so it can be found again. Do not fix or shorten it.
- "claim": the same statement rewritten to stand on its own (resolve "he", "it", "last year" into the names and dates the passage makes clear).

TEXT:
"""${text.slice(0, 9000)}"""

Reply with ONLY a JSON array of objects, no markdown fence:
[{"quote":"...","claim":"..."}]`;
}

function factCheckParseClaims(raw, limit) {
  const match = String(raw || "").match(/\[[\s\S]*\]/);
  if (!match) return [];
  let arr;
  try {
    arr = JSON.parse(match[0]);
  } catch (_) {
    return [];
  }
  if (!Array.isArray(arr)) return [];
  return arr
    .map((item) => {
      // Object form {quote, claim}; tolerate a bare string from older outputs.
      if (typeof item === "string") {
        const s = item.trim();
        return { claim: s, quote: s };
      }
      if (item && typeof item === "object") {
        const claim = String(item.claim || item.quote || "").trim();
        const quote = String(item.quote || item.claim || "").trim();
        return { claim, quote };
      }
      return null;
    })
    .filter((c) => c && c.claim.length > 12)
    .slice(0, limit);
}

// Extraction is ordinary text work on text we already hold, so the free
// on-device model is genuinely good enough — and keeping it here is what lets
// a whole-article check cost a handful of verdict calls instead of fifty.
async function factCheckExtractClaims(text, limit = 6) {
  const body = String(text || "").trim();
  if (body.length < 40) return [];
  const prompt = factCheckExtractPrompt(body, limit);

  const settings = await gatewayGetSettings();
  if (!settings) {
    // Same reasoning as factCheckVerify: without this, a keyless user gets an
    // empty claim list and a fact-check that appears to find nothing.
    const answer = await hostedComplete([{ role: "user", content: prompt }], { maxTokens: 700 });
    return factCheckParseClaims(answer, limit);
  }
  try {
    const res = await fetch("https://ai-gateway.vercel.sh/v1/chat/completions", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${settings.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: settings.model,
        messages: [{ role: "user", content: prompt }],
        max_tokens: 700,
        temperature: 0,
      }),
    });
    if (!res.ok) return [];
    const data = await res.json();
    return factCheckParseClaims(data?.choices?.[0]?.message?.content, limit);
  } catch (_) {
    return [];
  }
}

// ---------- Shared errata cache ----------
//
// A verdict belongs to the article, not to the reader who happened to trigger
// the check. So before spending anything, ask whether this exact revision has
// already been checked for somebody else. On a page anyone has read before this
// is the whole flow: no key, no model call, no wait.

const FC_ERRATA_READ = "https://api.justclarify.xyz/errata";
const FC_ERRATA_CHECK = "https://api.justclarify.xyz/errata/check";

// Must produce the same digest as errata.content_hash on the server, or the
// two sides will never agree on a cache key. Normalizing away whitespace and
// case first means a CMS re-wrapping a paragraph doesn't discard a good verdict.
async function factCheckContentHash(text) {
  const normalized = String(text || "").replace(/\s+/g, " ").trim().toLowerCase();
  const digest = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(normalized),
  );
  return Array.from(new Uint8Array(digest), (b) => b.toString(16).padStart(2, "0")).join("");
}

// Returns the cached verdicts, or null for any reason at all — miss, stale,
// server down, cache not configured. Every one of those means the same thing to
// the caller ("nothing cached, do it yourself"), so they collapse on purpose.
async function factCheckCacheRead(url, hash) {
  try {
    const res = await fetch(
      `${FC_ERRATA_READ}?url=${encodeURIComponent(url)}&content_hash=${hash}`,
    );
    if (!res.ok) return null;
    const data = await res.json();
    return data?.hit && Array.isArray(data.verdicts) && data.verdicts.length
      ? data.verdicts
      : null;
  } catch (_) {
    return null;
  }
}

// Cold article: have the server run the check on its own key and keep the
// result for the next reader. This is what lets someone with no API key check a
// page that nobody has read yet.
async function factCheckServerCheck(url, text, title) {
  try {
    const res = await fetch(FC_ERRATA_CHECK, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ url, text, title: title || "" }),
    });
    if (!res.ok) return null;
    const data = await res.json();
    return Array.isArray(data?.verdicts) && data.verdicts.length ? data.verdicts : null;
  } catch (_) {
    return null;
  }
}

// Push a finished set of verdicts to the page using the same message sequence a
// live run emits, so the content script needs no idea where they came from.
function factCheckDeliver(verdicts, tabId, runId) {
  const claims = verdicts.map((v) => ({
    claim: v.claim || v.quote || "",
    quote: v.quote || v.claim || "",
  }));

  if (tabId != null) {
    chrome.tabs.sendMessage(tabId, { type: "JC_FACTCHECK_CLAIMS", runId, claims }).catch(() => {});
    verdicts.forEach((result, index) => {
      chrome.tabs
        .sendMessage(tabId, { type: "JC_FACTCHECK_RESULT", runId, index, result })
        .catch(() => {});
    });
    chrome.tabs.sendMessage(tabId, { type: "JC_FACTCHECK_DONE", runId }).catch(() => {});
  }
  return { ok: true, claims, results: verdicts, cached: true };
}

// Check a body of text end to end: pull the claims worth checking, then verify
// them concurrently, reporting each verdict to the page as it lands so a long
// article fills in progressively instead of blocking on the slowest claim.
//
// `url` opts this run into the shared cache. Callers that pass none — live
// audio, where the "text" is a rolling transcript rather than a stable
// document — go straight to the local path, which is correct: a transcript
// fragment isn't an article and must never be stored against a page URL.
async function factCheckText(text, tabId, runId, limit = 6, url = "", title = "") {
  if (url) {
    const hash = await factCheckContentHash(text).catch(() => null);
    if (hash) {
      const cached = await factCheckCacheRead(url, hash);
      if (cached) return factCheckDeliver(cached, tabId, runId);

      // Nothing cached. Let the server check it — that keeps the cost one call
      // per article rather than one per reader, and fills the cache for
      // everyone after. Falls through to the local path if it can't.
      const computed = await factCheckServerCheck(url, text, title);
      if (computed) return factCheckDeliver(computed, tabId, runId);
    }
  }

  return factCheckTextLocal(text, tabId, runId, limit);
}

// The original path: extract and verify here, on the reader's own key. Still
// the fallback whenever the shared cache is unavailable, and still what live
// audio uses.
async function factCheckTextLocal(text, tabId, runId, limit = 6) {
  const claims = await factCheckExtractClaims(text, limit);
  if (!claims.length) {
    return { ok: true, claims: [], results: [] };
  }
  if (tabId != null) {
    chrome.tabs
      .sendMessage(tabId, { type: "JC_FACTCHECK_CLAIMS", runId, claims })
      .catch(() => {});
  }

  const results = await Promise.all(
    claims.map(async (claim, index) => {
      // Verify the self-contained form; the verdict inherits the claim's quote
      // so the page can tie the ruling back to the exact sentence it underlined.
      const result = await factCheckOne(claim.claim, text).catch(() => null);
      if (result) result.quote = claim.quote;
      if (result && tabId != null) {
        chrome.tabs
          .sendMessage(tabId, { type: "JC_FACTCHECK_RESULT", runId, index, result })
          .catch(() => {});
      }
      return result;
    }),
  );

  if (tabId != null) {
    chrome.tabs.sendMessage(tabId, { type: "JC_FACTCHECK_DONE", runId }).catch(() => {});
  }
  return { ok: true, claims, results: results.filter(Boolean) };
}
